specs/capri/MultiCloud_AK-SK_Management_Design.md

```diff
@@ -0,0 +1,74 @@
+
+    * Goal 2 - sub goal 2: Support signature validations for the inbound API request to multi-cloud.
+
+### Non Goals
+
+* Goal 1: Support role-based authorization for the multi-cloud users for API requests to PUT/POST/GET/DELETE a backend. Assign the set of roles the User has on multi-cloud.
+
+
+## Design Details
+
+The Multi-Cloud AK/SK Management design is shown in the following diagram:
+
+![Multi-Cloud AK-SK Management Diagram](multicloud_ak-sk_management.PNG?raw=true "Multi-Cloud AK-SK Management Diagram")
+
+Following are the steps for authenticating requests to Multi-Cloud API:
+
+1. Construct a request to API handler.
+
+2. Calculate the signature using signing key.
+
+3. Send the request (include access key ID and signature) to API handler.
+
+4. API Handler uses the access key ID to look up your secret access key from keystone.
+
+5. API Handler calculates a signature from the request data and the signing key using the same algorithm that user used to calculate the signature sent in the request.
+
+6. If the signature generated by API Handler matches with the user requested signature, the request is authentic. Else, the request is discarded and returns an error response.
+
+### AK/SK Credential Management
+
+User sends authentication credentials to keystone, the Identity service from keystone generates and returns a token. A token represents the authenticated identity of a user and, grants authorization. User can PUT/POST/GET/DELETE a credential as [here](https://developer.openstack.org/api-ref/identity/v3/index.html?expanded=list-credentials-detail,show-credential-details-detail,create-credential-detail#application-credentials).
+
+### Signature Calculation and Validation
+
+API requests to multi-cloud containing the authentication information must include a signature which is computed based on the request elements. To calculate a signature, concatenate request elements to form a string (string to sign) and then create a signing key using secret access key. Signing key is derived from the credential scope, which means that there is no need to include the key in the request. Then signing key is used to calculate the hash-based message authentication code (HMAC) of the string to sign.
+When API handler receives the request, it performs the same steps to calculate the signature included in the request. API handler then compares its calculated signature to the one sent with the request. If the signatures match, the request is processed. If the signatures don't match, the request is denied.
+
+AWS Signature Version 4 signing library i.e. aws-sigv4 can be used to generate the signature.
+
+Following is the illustration for the signature calculation and validation process.
+
+![Multi-Cloud Signature Calculation Diagram](multicloud_signature_calculation.PNG?raw=true "Multi-Cloud Signature Calculation Diagram")
```

Comment:
As discussed in the OpenSDS technical meeting, can you please make the text in this calculation diagram bigger so people can read it?

Reply:
Done